SI-7291: Don't throw exceptions while encountering diverging expansion.

hubertp opened the pull request:
Since we don't throw exceptions for normal errors it was a bit odd that we don't do that for DivergingImplicit. As SI-7291 shows, the logic behind catching/throwing exception was broken for divergence. Instead of patching it, I rewrote the mechanism so that we now another SearchFailure type related to diverging expansion, similar to ambiguous implicit scenario.
The logic to prevent diverging expansion from stopping the search had to be slightly adapted but works as usual.

The upside is that we don't have to catch diverging implicit for example in the presentation compiler which was again showing that something was utterly broken with the exception approach.

Review by @retronym.

hubertp committed:
SI-7291: No exception throwing for diverging implicit expansion

Since we don't throw exceptions for normal errors it was a bit odd
that we don't do that for DivergingImplicit.
As SI-7291 shows, the logic behind catching/throwing exception
was broken for divergence. Instead of patching it, I rewrote
the mechanism so that we now another SearchFailure type related
to diverging expansion, similar to ambiguous implicit scenario.
The logic to prevent diverging expansion from stopping the search
had to be slightly adapted but works as usual.

The upside is that we don't have to catch diverging implicit
for example in the presentation compiler which was again showing
that something was utterly broken with the exception approach.

paulp commented on src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala:
```diff
@@ -59,6 +59,25 @@ trait ContextErrors {
+      NormalTypeError(tree, errMsgForPt(pt), kind)
+    private def errMsgForPt(pt: Type) = 
+      "diverging implicit expansion for type "+pt+"\nstarting with "+
+      sym.fullLocationString
```
string interpolation is preferred.

paulp commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -820,11 +814,19 @@ trait Implicits {
+            if (search.isDivergent && divergenceCountdown > 0) {
+              divergenceCountdown -= 1
+              divergentSym = i.sym
+              log("discarding divergent implicit during implicit search")
```
How about "discarding divergent implicit $divergentSym".

paulp commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -783,16 +785,8 @@ trait Implicits {
-            SearchFailure
-        }
-      }
+      private var divergentSym: Symbol = null
```
Can you comment that you are using null as a marker?

retronym commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -820,11 +814,19 @@ trait Implicits {
+              divergenceCountdown -= 1
+              divergentSym = i.sym
+              log("discarding divergent implicit during implicit search")
+              SearchFailure
```
Might be nicer to encapsulate the mutable fields for countdown/recovery logic.

``` scala
object DivergentImplicitRecovery {
  private var sym: Symbol = _
  private var countdown: Int = 1
  def apply(search: SearchResult): SearchResult = ...
}

...

val search = typedImplicit(i, ptChecked = true, isLocal)
DivergentImplicitRecovery(search)
```

retronym commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -151,14 +151,14 @@ trait Typers extends Adaptations with Tags {
             if (!param.hasDefault && !paramFailed) {
               context.errBuffer.find(_.kind == ErrorKinds.Divergent) match {
-                case Some(divergentImplicit) =>
+                case Some(divergent: DivergentImplicitTypeError) =>
```
Seems a bit redundant to have both `ErrorKind.Divergent` and the type `DivergentImplicitTypeError`. Maybe we can do with one of them, and have a `collect` method for errors.

hubertp replied:
Yeah, I did think about it but then a) we would have different behaviour between lazy SearchFailures b) we still do have to provide more specialized parameter type (see a couple of lines below) anyway for better error messages and unfortunately one cannot do it earlier c) here we handle the "no found implicit" and "diverging expansion..." but there are situations where we don't do it and simply rely on the fact that buffering the errors will do the right job.
I can definitely agree that the situation is not ideal but then choosing between any of the two methods only is also inconvenient.  Let me see if I can simplify it. I am not sure I get your `collect` proposition, could you elaborate?

retronym replied:
``` scala
val x: Option[DTE] = context.errBuffer.collectFirst { 
  case dte: DivergentImplicitTypeError => dte
}
```

Basically, should we get rid of `ErrKind` and just use the type of the error? But we could (should?) do this on a followup change, to help me get my `Context` refactoring in.

hubertp replied:
It's tricky. I tried to change it by basically relying only on (Diverging)SearchResult but the issue becomes with having precise error message - the divergent symbol is lost by the time we want to finally report the error message. The only way to recover it is through the context buffer.

paulp commented:
Here is a more efficient version of normalizeFirstParent. It avoids duplicating the isTraitRef test, and doesn't create new cons cells unless they are needed. (When ys == x :: xs and you return x :: xs instead of ys, you allocate unnecessarily.) This is untested, even to the extent of compiling it, but I'm unreasonably optimistic it will work.

```
    private def normalizeFirstParent(parents: List[Tree]): List[Tree] = {
      @annotation.tailrec def explode(parents: List[Tree]): List[Tree] = {
        val supertpt :: rest = parents // parents is always non-empty here - it only grows
        if (treeInfo isTraitRef supertpt) {
          val supertpt1  = typedType(supertpt)
          def supersuper = TypeTree(supertpt1.tpe.firstParent) setPos supertpt.pos.focus
          if (supertpt1.isErrorTyped) rest
          else explode(supersuper :: supertpt1 :: rest)
        }
        else {
          if (supertpt.tpe.typeSymbol == AnyClass)
            supertpt setType AnyRefTpe

          parents
        }
      }
      if (parents.isEmpty) Nil else explode(parents)
    }
```

hubertp commented:
Sorry to kill your optimism but this version won't work with value classes (that's the reason why there was a duplicate isTraitRef test). Allocation hint is still useful, though. Thanks.

paulp commented:
@hubertp Can you explain? Why doesn't it work for value classes? I am confident whatever the reason is, I can modify this code to cover it without duplicating the test.

hubertp committed:
SI-7291: Remove error kinds.

What seemed like a good idea initially (since potentially there were
many different kinds of errors that could be treated specially),
started to complicate the error logic.
So let's just match on the specific instance of an error to
manipulate the buffer.

paulp commented:
Also, please limit commit message line lengths to 72 characters, as noted (with inadequate emphasis) in the pull request policy link to http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html . 364 slightly exceeds the limit.

hubertp commented:
@paulp I should have been more precise. There are bootstrapping issues with AnyVal in that version. I didn't mean it wasn't possible at all, just that there were problems with the version you proposed. Let me tinker it a little bit more.

Updated wrt to comments, rebased against current master that has the new context impl.

paulp commented:
@hubertp I tried it out and did encounter the discussed bootstrapping problems. I'm a little confused about why, but the fact that there's a genuine obstacle is enough for me. This LGTM then.

paulp commented:
I'm going to merge this because it sounds like it will save #2406.

xeno-by commented:
May I port this to 2.10.x?

paulp commented:
No problem that I can see.

hubertp commented:
Were Jason's context changes backported to 2.10? That can be problematic for you, if not.

paulp commented:
@hubertp I'm curious what is the story here: https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos-per-commit/1949/consoleText

It fails the same way as before ("ERROR: Right(scala.tools.nsc.typechecker.DivergentImplicit$)") and it includes this PR. I got the impression one or more someones thought it wouldn't happen like that.

hubertp commented:
@paulp I guess I am seeing ghosts. DivergentImplicit is gone, seriously. Let me have a closer look.
